platforms/core-configuration/declarative-dsl-provider/src/integTest/groovy/org/gradle/internal/declarativedsl/settings/SoftwareTypeConventionIntegrationTest.groovy

```diff
@@ -0,0 +1,127 @@
+        "no conventions, id is set"            | ""                                   | setId("test")         | """id = test\nbar = bar"""
+        "everything has convention and is set" | setAll("convention", "convention")   | setAll("test", "baz") | """id = test\nbar = baz"""
+    }
+
+    def "can configure build-level conventions in a non-declarative settings file"() {
+        given:
+        withSoftwareTypePlugins().prepareToExecute()
+
+        file("settings.gradle.kts") << getDeclarativeSettingsScriptThatSetsConventions(setAll("convention", "convention"))
+
+        file("build.gradle.dcl") << getDeclarativeScriptThatConfiguresOnlyTestSoftwareType(setId("test"))
+
+        when:
+        run(":printTestSoftwareTypeExtensionConfiguration")
+
+        then:
+        outputContains("""id = test\nbar = convention""")
+    }
+
+    def "can configure build-level conventions in a declarative settings file and apply in a non-declarative project file"() {
+        given:
+        withSoftwareTypePlugins().prepareToExecute()
+
+        file("settings.gradle.dcl") << getDeclarativeSettingsScriptThatSetsConventions(setAll("convention", "convention"))
+
+        file("build.gradle.kts") << """
+            plugins { id("com.example.test-software-type-impl") }
+        """ + getDeclarativeScriptThatConfiguresOnlyTestSoftwareType()
+
+        when:
+        run(":printTestSoftwareTypeExtensionConfiguration")
+
+        then:
+        outputContains("""id = convention\nbar = convention""")
+    }
+
+    def "can configure build-level conventions in a settings plugin"() {
+        given:
+        withSettingsPluginThatConfiguresSoftwareTypeConventions().prepareToExecute()
+
+        file("settings.gradle.dcl") << getDeclarativeSettingsScriptThatSetsConventions()
+
+        file("build.gradle.dcl") << getDeclarativeScriptThatConfiguresOnlyTestSoftwareType(setId("test"))
+
+        when:
+        run(":printTestSoftwareTypeExtensionConfiguration")
+
+        then:
+        outputContains("""id = test\nbar = plugin""")
+    }
+
+    static String setId(String id) {
+        return "id = \"${id}\""
+    }
+
+    static String setFooBar(String bar) {
+        return "foo { bar = \"${bar}\" }"
+    }
+
+    static String setAll(String id, String bar) {
+        return setId(id) + "\n" + setFooBar(bar)
+    }
+
+    static String getDeclarativeScriptThatConfiguresOnlyTestSoftwareType(String configuration="") {
+        return """
+            testSoftwareType {
+                ${configuration}
+            }
+        """
+    }
+
+    static String getDeclarativeSettingsScriptThatSetsConventions(String configuration="") {
+        return """
+            pluginManagement {
+                includeBuild("plugins")
+            }
+            plugins {
+                id("com.example.test-software-type")
+            }
+
+            testSoftwareType {
```

Comment:
I wonder if it would be clearer if the software types would be exposed not in the immediate scope of `Settings` but in a nested scope with a clearer name like `conventions {...}`. But that's not something we should finalize right now, anyway.

Reply:
I don't feel strongly about this either way.  I had originally thought I would add a top-level `conventions {}` block, but then decided it wasn't adding any value beyond simply nesting the configuration and thought it might be simpler for users if the convention blocks were at the top-level just like they are in the project scripts.  I could easily be convinced to move it into a nested block, though.  I'd be interested if anyone else has any strong opinions about this...